### ISSUE TITLE
Reason for Generic Access Rights in CreateFile Callback?

### ISSUE THREAD

**Issue by lumip** [dokan-dev/dokan-dotnet#194]
We noticed that the generic access flags (GENERIC_READ, GENERIC_WRITE, etc) are propagated directly into our user mode implementation via the DokanOperations.CreateFile callback, i.e., a conversion into file specific access rights is not performed.

Looking at the code, it seems like the conversion actually does happen in the kernel driver/library but is reverted in the .NET-wrapper in DokanOperationProxy.ZwCreateFileProxy https://github.com/dokan-dev/dokan-dotnet/blob/85165d5da69f51b9b6ae0df5e2ce35d8e0a1bb51/DokanNet/DokanOperationProxy.cs#L293-L300

Is there a specific reason for this?

**Comment by Liryna** [comment]
Hi @lumip ,

Yes, there is an obscure logic reason for this 😄 
See https://github.com/dokan-dev/dokany/issues/197#issuecomment-250967638
https://github.com/dokan-dev/dokany/blob/0da77870161c101d3133fd76938951c200f12868/dokan/dokan.h#L844

Do not hesitate if you need more information.

**Comment by lumip** [comment]
hm.. there are several things I don't really understand here:

1. How does the simple act of re-introducing the Generic Access Rights _towards the user mode callbacks_ fix the bug in that discussion?

2. Why does this back-conversion only happen in the .NET-wrapper when the linked discussion is actually part of the core dokany repository?

**Comment by Liryna** [comment]
1. Kernel add/convert additional flags before forwarding it to the driver. These flags used as it is directly to the userland created an invalid parameters error.

2. The "issue" is that Kernel CreateFile is more advanced than the userland CreateFile, there are options that only exist at this point and "some" users would probably want/need to get the full information. Therefore we chose not to convert them and let the user implementation convert the flag with a helper.
This function is used in the C/C# Mirror.

**Comment by lumip** [comment]
1. That's about as far as I understood it, but thanks for summing it up again :) It just makes me wonder why that is since the Windows documentation is pretty explicit about that the generic access should be mapped to specific access for file objects. To me, this suggests that a file system driver implementation should be pretty agnostic about generic access rights.. Basically I don't understand how this solution fixes the bug described in the conversation you linked (and apparently I'm not the only one: https://github.com/dokan-dev/dokany/issues/197#issuecomment-251147310 )

2. I'd believe that most userland code does not need the generic rights and would need to map them again to the specific ones, making a total of three conversions (generic -> specific in the driver, specific -> generic in the .NET-wrapper, generic -> specific in the user application).. in that case I would argue not to do the re-mapping in the .NET-wrapper by default but expose the conversion function (`DokanMapKernelToUserCreateFileFlags`) in the interface so that those who actually do need can then use it. I agree with the other conversions that take place in the mapping function (disposition, attributes, ..) but really take issue with the desired access mask mapping.

Also: what if the initial call to `CreateFile` (to the driver) had a desired access of `FILE_READ_ATTRIBUTES | FILE_READ_DATA | FILE_READ_EA | STANDARD_RIGHTS_READ | SYNCHRONIZE` (i.e., everything that would constitute `GENERIC_READ` but not explicitely being `GENERIC_READ`)? The `DokanMapKernelToUserCreateFileFlags` would then map this to `GENERIC_READ` and pass this into the userland file system callbacks even though this is *not* what was given in the original call to `CreateFile`...

**Comment by Liryna** [comment]
1. Example (not tested, just from my mind):
- User app call CreateFile `GENERIC_READ`
- Kernel convert the `GENERIC_READ` into `STANDARD_RIGHTS_READ, FILE_READ_DATA, FILE_READ_ATTRIBUTES, FILE_READ_EA, and SYNCHRONIZE. `
- Kernel forward to driver `STANDARD_RIGHTS_READ, FILE_READ_DATA, FILE_READ_ATTRIBUTES, FILE_READ_EA, and SYNCHRONIZE. ` not `GENERIC_READ`
- Dokan userland gets `STANDARD_RIGHTS_READ, FILE_READ_DATA, FILE_READ_ATTRIBUTES, FILE_READ_EA, and SYNCHRONIZE. ` And call `CreateFile` directly with it. This creates an invalid parameters.

The conversion back fixes the invalid parameters during the `CreateFile` in userland.
https://github.com/dokan-dev/dokany/blob/af8ddfc372246b01832ea37924ef513ddc3ed299/dokan/dokan.c#L914

2. The goal of the C# wrapper is simplicity that’s why we preferred to make the conversion by default in the proxy. More over, it is to give to the user a clean `FileAccess` (framework available flags). If you would like to get all the values of the createfile in c#, would probably need to find a way to by pass the `ZwCreateFileProxy`.

https://github.com/dokan-dev/dokany/blob/af8ddfc372246b01832ea37924ef513ddc3ed299/dokan/dokan.c#L916
See that `FILE_GENERIC_READ` is used and not `GENERIC_READ`

**Comment by lumip** [comment]
> Dokan userland gets STANDARD_RIGHTS_READ, FILE_READ_DATA, FILE_READ_ATTRIBUTES, FILE_READ_EA, and SYNCHRONIZE. And call CreateFile directly with it. This creates an invalid parameters.

Is that explicetly a problem for the mirror sample implementation? Why would those be invalid parameters in general?

> The goal of the C# wrapper is simplicity that’s why we preferred to make the conversion by default in the proxy. More over, it is to give to the user a clean FileAccess (framework available flags). If you would like to get all the values of the createfile in c#, would probably need to find a way to by pass the ZwCreateFileProxy.

I'm confused by this.. FileAccess would be cleaner without the generic rights (as they are duplicating specific rights inconveniently). In my opinion, this conversion in the wrapper actually decreases simplicity for the programmer of the callbacks (and the wrapper code itself).

> https://github.com/dokan-dev/dokany/blob/af8ddfc372246b01832ea37924ef513ddc3ed299/dokan/dokan.c#L916
See that FILE_GENERIC_READ is used and not GENERIC_READ

The next line ( https://github.com/dokan-dev/dokany/blob/af8ddfc372246b01832ea37924ef513ddc3ed299/dokan/dokan.c#L917 ) sets `GENERIC_READ` and another line ( https://github.com/dokan-dev/dokany/blob/af8ddfc372246b01832ea37924ef513ddc3ed299/dokan/dokan.c#L933 ) removes all attributes that constitute `GENERIC_READ` (which are `or`ed together as `FILE_GENERIC_READ`)

I'm sorry, I don't want to be contrary but I still don't understand why this is necessary or even beneficial. Please enlighten me :)

**Comment by Liryna** [comment]
1. See my comment here https://github.com/dokan-dev/dokany/issues/197#issuecomment-250535936 both flag was set and created the invalid parameters. Yes this is an issue for the mirror because we use the userland CreateFile that create this error. if you do not use it and have no issue with managing this flag, you can not to use the convert function.

2. C# Access rights https://msdn.microsoft.com/fr-fr/library/4z36sx0f(v=vs.110).aspx
C CreateFile https://github.com/dokan-dev/dokany/blob/master/samples/dokan_mirror/mirror.c#L243-L262
C# is an enum compared to C. Having a clean enum variable means only existing enum value is set.
What would be the point as giving corrupt enum to the userland ?

3. 
```
#define FILE_GENERIC_READ         (STANDARD_RIGHTS_READ     |\ //0x00020000L
                                   FILE_READ_DATA           |\ //0x0001
                                   FILE_READ_ATTRIBUTES     |\ //0x0080
                                   FILE_READ_EA             |\ //0x0008
                                   SYNCHRONIZE) //0x00100000L
#total = 0x00120089
#define GENERIC_READ                     (0x80000000L)
```
Both are fully different flags.

**Comment by lumip** [comment]
1. Shouldn't this be fixed in the mirror then instead of the wrapper?
> If you do not use it and have no issue with managing this flag, you can not to use the convert function.

currently, I do not have this choice because the conversion happens automatically in the wrapper..

2. I don't want to give a corrupt enum to the user callback. Just one without generic access rights :)

3. Yea, the are represented by different values but semantically they are the same and thus I'd say to remove the GENERIC_{READ|WRITE|EXECUTE|ALL} (in the DokanNet.FileAccess enum and thus the call into DokanOperations.CreateFile) to avoid more complicated/duplicated handling routines in that callback.


but.. I'll stop here. I see you made your choices about this, even though I can't really comprehend them. It's not too bad, it just seems unclean to me so I wanted to suggest changing that. I guess that's not going to happen.
Thanks for taking the time for this discussion!

**Comment by Liryna** [comment]
1. Like I said, the C# wrapper is not made to be a low level as the C. If you would like to control this part, you will need to change the proxy not to call the `DokanMapKernelToUserCreateFileFlags` or to direct C that does not call this function by default.

2. So here is the case 👍 The conversion is made to give the proper generic access rights.

3. Yes, I understand the point here. I would totally agree with you.

But really, I have looked for a long time why CreateFIle would create an invalid parameters error during VHD mount. I have reproduced the param calls on a sample. Eliminated all the parameters one by one to find the conflict. I see which one and tried why the kernel would ask incompatible createfile params and then I found the previous link that explains the kernel conversion. I wrote the function to convert back and it worked directly, had no issue with other tests to it felt to me to be the right fix. It is rare in case such to find a Microsoft documentation that can be linked to an apply fix so I am pretty happy of it 😄 

If you find that it is not the correct way to do it, I would be glad to see another solution. I am totally open since this comes only from my own thought.

Just a point, I do not take bad your words if you feel it like this, I am just having difficult to explain this fix/situation where I cannot point a code/documentation that tells, "how to convert desired access kernel to users".
